internal/service/cluster/cluster_e2e_test.go

```diff
@@ -36,48 +37,62 @@ import (
 )
 
 type testCase struct {
-	name                 string
-	nodeCountInitial     int
-	initialIsolatedNodes []string // list of node names that are initially in isolated network space
-	assertionsInitial    func(t *assert.CollectT, state *testState)
-	changes              func(state *testState)
-	assertionsFinal      func(t *assert.CollectT, state *testState)
-	assertionsTimeout    time.Duration
-	extraAllowedErrors   []string
-	alloyConfig          string
+	name                   string
+	nodeCountInitial       int
+	initialIsolatedNodes   []string // list of node names that are initially in isolated network space
+	assertionsInitial      func(t *assert.CollectT, state *testState)
+	changes                func(state *testState)
+	assertionsFinal        func(t *assert.CollectT, state *testState)
+	assertionsTimeout      time.Duration
+	extraAllowedErrors     []string
+	alloyConfig            string
+	minimumClusterSize     int
+	minimumSizeWaitTimeout time.Duration
 }
 
 func TestClusterE2E(t *testing.T) {
 	tests := []testCase{
 		{
 			name:             "three nodes just join",
+			alloyConfig:      `testcomponents.cluster_state_tracker "foo" {}`,
 			nodeCountInitial: 3,
 			assertionsInitial: func(t *assert.CollectT, state *testState) {
 				for _, p := range state.peers {
 					verifyMetrics(t, p,
 						`cluster_node_info{state="participant"} 1`,
 						`cluster_node_peers{cluster_name="cluster_e2e_test",state="participant"} 3`,
 						`cluster_node_gossip_alive_peers{cluster_name="cluster_e2e_test"} 3`,
+						`cluster_state{component_id="testcomponents.cluster_state_tracker.foo",component_path="/",peers="node-0___node-1___node-2",ready="true"} 3`,
 					)
 					verifyPeers(t, p, 3)
+					verifyClusterReady(t, p)
 				}
 				verifyLookupInvariants(t, state.peers)
 			},
 		},
 		{
 			name:             "three nodes with slow components",
 			nodeCountInitial: 3,
+			// slow components can currently lead to timeouts and communication errors
+			// TODO: consider decoupling cluster operations from runtime/components performance
```

Comment:
I actually have done this by putting the dispatching of cluster updates on a separate goroutine. TODO: remove this comment.